### PR TITLE
Client: Terminal block validation

### PR DESCRIPTION
**Pull request by cbrzn**
This PR aims to add terminal block validation in `newPayload` and `forkchoiceUpdated` methods from Engine API. The reason is that we need to add an explicit check in these methods to return the expected response based on the [specs](https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#payload-validation)

I ran the tests against the hive and two of them went green :-D

### CODE REVIEW

File: packages/client/lib/rpc/modules/engine.ts [hunk 1]
```diff
@@ -173,6 +173,25 @@ const validHash = async (
...
 }
 
+/**
+ *  Validate that the head block sent is a not a PoW block
+ */
+const validateTerminalBlock = async (block: Block, chain: Chain): Promise<boolean> => {
+  const ttd = chain.config.chainCommon.hardforkTD()
+  if (ttd === null) return false
+  const blockTd = await chain.getTd(block.hash(), block.header.number)
+  if (!block.header.parentHash.toString()) {
```
Comment (ryanio) [inline packages/client/lib/rpc/modules/engine.ts:183]: awesome PR, nice work :) what's this if condition checking?

File: packages/client/test/rpc/engine/forkchoiceUpdatedV1.spec.ts [hunk 2]
```diff
@@ -93,6 +93,26 @@ tape(`${method}: call with valid fork choice state without payload attributes`,
...
+    },
+  }
+
+  const { server } = await setupChain(genesisWithHigherTtd, 'post-merge', {
+    engine: true,
+  })
+
+  const req = params(method, [validForkChoiceState, null])
+  const expectRes = (res: any) => {
+    t.equal(res.body.result.payloadStatus.status, 'INVALID_TERMINAL_BLOCK')
```
Comment (ryanio) [inline packages/client/test/rpc/engine/forkchoiceUpdatedV1.spec.ts:111]: thanks for adding a test! can you also add one to newPayload so we can get coverage on the error added that as well

File: packages/client/lib/rpc/modules/engine.ts (hunk 1 above)
Comment (cbrzn) [reply packages/client/lib/rpc/modules/engine.ts:183]: It checks if the parent hash it's not `0x0..0` - This is not happening in mainnet but I thought that it's more for development and experimental purposes?

File: packages/client/test/rpc/engine/forkchoiceUpdatedV1.spec.ts (hunk 2 above)
Comment (cbrzn) [reply packages/client/test/rpc/engine/forkchoiceUpdatedV1.spec.ts:111]: done :-D

File: packages/client/lib/rpc/modules/engine.ts (hunk 1 above)
Comment (ryanio) [reply packages/client/lib/rpc/modules/engine.ts:183]: hm that would be `block.header.parentHash.equals(Address.zero())`

maybe add a comment under the if statement that says `// This should only happen for development purposes, for when the genesis block has parentHash 0x000...000`

you could also probably also do `!block.isGenesis()`

Comment (g11tech) [reply packages/client/lib/rpc/modules/engine.ts:183]: yes, for (dev) chains starting from a genesis block >=TTD, genesis block is the terminal block, the parentHash should be compared against zero hex, or as Ryan suggested !block.isGenesis

File: packages/client/lib/rpc/modules/engine.ts [hunk 3]
```diff
@@ -291,7 +310,15 @@ export class Engine {
     const common = this.config.chainCommon
 
     try {
-      await findBlock(toBuffer(parentHash), this.validBlocks, this.chain)
+      const block = await findBlock(toBuffer(parentHash), this.validBlocks, this.chain)
+      const validTerminalBlock = await validateTerminalBlock(block, this.chain)
```
Comment (g11tech) [inline packages/client/lib/rpc/modules/engine.ts:314]: I think this check should only be done if this parent block is "pre-merge", else this check will fail for a post-merge block

File: packages/client/lib/rpc/modules/engine.ts [hunk 4]
```diff
@@ -429,6 +457,18 @@ export class Engine {
         return { payloadStatus, payloadId: null }
       }
     }
+    const validTerminalBlock = await validateTerminalBlock(headBlock, this.chain)
```
Comment (g11tech) [inline packages/client/lib/rpc/modules/engine.ts:460]: again this check should be done if headblock is a "pre-merge" block, check will fail on a post merge block

**Comment by holgerd77** [comment]
Guys, just to let you know: I will leave this out of the current release round, if this is still so actively in review, guess we very much need a cut now (so please also wait with merging until releases from #1774 are done).

We can very well do some bugfix releases on 1-3 libraries next week as some follow-up on this.

**Comment by g11tech** [comment]
@cbrzn Can you retest this against hive test vectors, am testing this against the kiln network if everything goes :heavy_check_mark:

**Review by g11tech — DISMISSED** [review]
Ran successfully post merge on devnet 5, and pre-merge to post-merge transition on kiln! :rocket:  :partying_face:

**Comment by cbrzn** [comment]
> @cbrzn Can you retest this against hive test vectors, am testing this against the kiln network if everything goes heavy_check_mark

of course! thanks for helping here ser :pray:

File: packages/client/lib/rpc/modules/engine.ts [hunk 5]
```diff
@@ -173,6 +175,27 @@ const validHash = async (
   return bufferToHex(hash)
 }
 
+/**
+ *  Validate that the head block sent is a not a PoW block
```
Comment (ryanio) [inline packages/client/lib/rpc/modules/engine.ts:179]: could be PoA or PoW

```suggestion
 *  Validate that the block satisfies post-merge conditions.
```

File: packages/client/lib/rpc/modules/engine.ts [hunk 6]
```diff
@@ -173,6 +175,27 @@ const validHash = async (
...
+ */
+const validateTerminalBlock = async (block: Block, chain: Chain): Promise<boolean> => {
+  const ttd = chain.config.chainCommon.hardforkTD()
+  if (ttd === null) return false
+  const blockTd = await chain.getTd(block.hash(), block.header.number)
+  // Block is terminal if its td >= ttd and and its parent td < ttd. In case the Genesis block
+  // has td >= ttd it is the terminal block
+  if (!block.isGenesis()) {
```
[the remark this answers is not in the document]
Comment (ryanio) [reply packages/client/lib/rpc/modules/engine.ts:187]: codecov is also saying this path is uncovered, could you add one test for this too? everything else looks great!

[the remark this answers is not in the document]
Comment (cbrzn) [reply packages/client/lib/rpc/modules/engine.ts:187]: I thought this too... Will apply that

File: packages/client/lib/rpc/modules/engine.ts [hunk 7]
```diff
@@ -300,7 +323,17 @@ export class Engine {
     const common = this.config.chainCommon
 
     try {
-      await findBlock(toBuffer(parentHash), this.validBlocks, this.chain)
+      const block = await findBlock(toBuffer(parentHash), this.validBlocks, this.chain)
+      if (!block._common.hardforkGteHardfork(block._common.hardfork(), Hardfork.Merge)) {
```
Comment (ryanio) [inline packages/client/lib/rpc/modules/engine.ts:327]:
```suggestion
      if (!block._common.gteHardfork(Hardfork.Merge)) {
```

**Review by g11tech — DISMISSED** [review]
LGTM! :heart:

**Review by ryanio — CHANGES_REQUESTED** [review]
@cbrzn and i discussed a few last items and he should follow with a commit once he can :)

**Review by ryanio — APPROVED** [review]
lgtm! great PR! can merge after ci passes

**Pull request merged**